### ISSUE TITLE
Student can 'Edit' and 'Delete' question of others after remove capability 'previewother'

### ISSUE THREAD

**Issue by HuongNV13** [frankkoch/moodle-mod_studentquiz#249]
1. Login as Admin/Manager
2. Go to StudentQuiz.
3. Create question (True/False).
4. Create student account.
5. Allow the "previewother" capability for that student account.
6. Log in as student.
7. Go to StudentQuiz above and verify that student can preview/edit/delete the question.
8. Login back to admin.
9. Disallow the "previewother" capability for that student account.
6. Log in as student again and go to StudentQuiz above
7. Verify that student cannot preview/edit/delete the question.

Actual: Student still can see the edit and delete button

**Comment by dionysius** [comment]
You can see the actual permissions by role if you go to the StudentQuiz -> gear icon -> check permissions. Keep in mind that those permissions are only in that context if you added those permissions in the context!

After adding and removing the previewothers permission we are at:

![image](https://user-images.githubusercontent.com/1341084/84713207-caadbf80-af6a-11ea-809d-7b43b450e42e.png)
![image](https://user-images.githubusercontent.com/1341084/84713226-d7321800-af6a-11ea-8046-1b5e82e3fe48.png)

As we can see, the critical permissions in questions got automatically added. I also think those should also not be added at all! This is probably due to https://github.com/frankkoch/moodle-mod_studentquiz/blob/98252597b7b42e4c90bd0c187a9c65f33ea6273f/locallib.php#L1006.

Independent of what capabilities are set, the more important question is also: how to get the "inherited" capabilities automatically removed? It looks like removing a studentquiz capability does not remove the "inherited" capability

**Comment by dionysius** [comment]
If a student can preview others, he should definitely not see the option to change the state **and not be able to actually change the state!**

![image](https://user-images.githubusercontent.com/1341084/84714283-87088500-af6d-11ea-90a7-1a20ca30fa23.png)

**Comment by dionysius** [comment]
I started a branch permission and I found an additional thing to discuss about (See current diff: https://github.com/frankkoch/moodle-mod_studentquiz/compare/permissions)

previewothers capability should IMO not allow for changing states. It should only allow previewing other questions. That leads to the next question: Changing a state of a question is something that non-editing teachers would do, while "manage" is the perfect capability for editing teachers, how should we assign them a capability to do that?

**Comment by dionysius** [comment]
I propose the following:

- preview.php needs a require_capability previewothers
- introduce a new capability changestate
- displaying the state change needs a check to has_capability changestate (in preview and in attempt, but they anyway use the same function for that)
- saving the state change needs a check to require_capability changestate
- remove moodle/question:view* dependencies everywhere, see [1]
- if a permission is removed / not granted, the dependant moodle/question:* permissions need to be removed automatically as well

[1] moodle/question:view* are obsolete: they allow users to go to the edit page of the question, but cannot save it. Preview and attempt require only moodle/question:use*

@timhunt There's something strange with moodle/question:tag*, they are not visible in the activity context (see https://github.com/frankkoch/moodle-mod_studentquiz/issues/249#issuecomment-644431258). I've found them in the code ([A](https://github.com/moodle/moodle/search?q=%22moodle%2Fquestion%3Atagmine%22&unscoped_q=%22moodle%2Fquestion%3Atagmine%22), [B](https://github.com/moodle/moodle/search?q=%22question_has_capability_on%28%24question%2C+%27tag%27%29%22&unscoped_q=%22question_has_capability_on%28%24question%2C+%27tag%27%29%22)), is there a reason why this is not listed in the acitivity context or is that a bug?

Rule of thumb:
Yes we can do studentquiz capabilities as we please, but we need to make sure they reflect the question capabilities so no unauthorized changes can be constructed

**Comment by dionysius** [comment]
Do we want to finegrain the capability manage? Currently manage allows the user to:

- edit and delete any question (via moodle/question:editall)
- move questions into a subcategory (*not explicitly defined*, so course capabilities are in effect - moodle/question:moveall)

We could make a capability mod/studentuiz:organize which sets dependency moodle/question:moveall, which allows someone to move questions around while he has not the manage capability. So manage means only edit and delete questions.

**Comment by timhunt** [comment]
Well, view* bapabilties have a particular special use which is not currently relevant to StudentQuiz, but it has its place. (Actually, it might be quite useful for StudentQuiz. The point is that it lets someone open the Edit question form read-only. So, if a student previews anther students' question, and things "That's cool, I wonder how they made that", then this capability would let them see, without being able to edit the other student's question.)

The reason you cannot see the moodle/question:tag capability in the Moodle permissions UI for activities is a simple bug. Those capabilites are not in the list returned by question_get_question_capabilities (defined in lib/questionlib.php). Please can you create a Moodle tracker issue for this (https://tracker.moodle.org/). Do you want to develop the fix, or shall I?

**Comment by dionysius** [comment]
I'll report the issue to the tracker, I'm not sure how i contribute to moodle itself, but I can give you a little patch there as well.

**Comment by timhunt** [comment]
Don't worry. If you don't want ot do it, I can do it easily enough.

**Comment by dionysius** [comment]
I'll setup the issue, since I can give screenshots and have found out additional stuff missing

**Comment by dionysius** [comment]
[MDL-69077](https://tracker.moodle.org/browse/MDL-69077)

**Comment by dionysius** [comment]
> So, if a student previews anther students' question, and things "That's cool, I wonder how they made that", then this capability would let them see, without being able to edit the other student's question.)

Agree, I propose adding viewall as dependency to previewothers, so a preview also gives you permission to see the question edit page read-only

**Comment by HuongNV13** [comment]
> 
- preview.php needs a require_capability previewothers (or it's his own question)
- introduce a new capability changestate
- displaying the state change needs a check to has_capability changestate (in preview and in attempt, but they anyway use the same function for that)
- saving the state change needs a check to require_capability changestate
- remove moodle/question:view* dependencies everywhere, see [1]
- if a permission is removed / not granted, the dependant moodle/question:* permissions need to be removed automatically as well

I agree with this solution. The hard part here is we need to check that if the permission were removed, we need to remove the relevant permissions related to moodle/question also.

> So, if a student previews anther students' question, and things "That's cool, I wonder how they made that", then this capability would let them see, without being able to edit the other student's question.)

Also, this is a very good idea. So can we still show the edit button but it will display the read-only form?

**Comment by timhunt** [comment]
I think that at the moment, our solution to the 'hard part' is the function mod_studentquiz_ensure_question_capabilities. That is supposed to veryify that, whatever combination of studentquiz capabilities you have, you have the right combination of question capabilities. I think this is good. We just need to make sure we call this function whenever necessary (without calling it so much that we get performance problems!)

Regarding the view capability, core question bank uses a different icon when the form will be read-only: https://github.com/moodle/moodle/blob/a2b286ce176fbe361f0889abc8f30f043cd664ae/question/classes/bank/edit_action_column.php#L57. (Not sure that is the best choice of icon though.)

**Comment by dionysius** [comment]
> Regarding the view capability, core question bank uses a different icon when the form will be read-only
Yeah I saw that too

> Not sure that is the best choice of icon though
I don't like the icon (gear) for editing the question, there are better icons for expressing "edit".

> the 'hard part' is the function mod_studentquiz_ensure_question_capabilities
If I remember correctly that part was also hard in the past in the upgrade.php to make sure, we are at the state where we should be

[1] So, If we define the capabilities like defined https://github.com/frankkoch/moodle-mod_studentquiz/issues/249#issuecomment-645299829, we have two approaches for display:
- Use the core question column, so the icons for view is "i", edit is "gear" and delete is "trashcan"
- Make our own question colum (as already proposed in #253) and allows us to use our own capabilities and custom icons

I'll push to branch permissions shortly with how we discussed the dependency releations here, so only open then are:

- decide variant for [1] and implement
- find a way to delete permissions, but not too often as said in https://github.com/frankkoch/moodle-mod_studentquiz/issues/249#issuecomment-645925019
- check and implement steps to upgrade.php if we need to be careful when upgrading

I'm on holiday from saturday on for a week, so I can't work on that for a little while. I think this issue is important to address soon as we expose currently more permissions (luckily *only* in the context of that activity instance)

**Comment by dionysius** [comment]
The permissions branch is ready to be used as base for further implementations: https://github.com/frankkoch/moodle-mod_studentquiz/compare/permissions
